src/main/scala/net/snowflake/spark/snowflake/ServerConnection.scala

```diff
@@ -0,0 +1,290 @@
+package net.snowflake.spark.snowflake
+
+import net.snowflake.client.core.{SFBaseSession, SFSession, SFSessionProperty}
+import net.snowflake.client.jdbc.SnowflakeConnectionV1
+import net.snowflake.spark.snowflake.Parameters.MergedParameters
+import net.snowflake.spark.snowflake.Utils.JDBC_DRIVER
+import org.apache.spark.sql.execution.datasources.jdbc.DriverRegistry
+import org.slf4j.LoggerFactory
+
+import java.sql.{Connection, DriverManager, PreparedStatement, SQLException, Statement}
+import java.util.Properties
+import java.util.concurrent.atomic.AtomicLong
+import scala.collection.mutable
+
+/*
+ * This class is the key if a JDBC connection is cached for sharing.
+ */
+private[snowflake]
+sealed class ConnectionCacheKey(private val parameters: MergedParameters) {
+  ServerConnection.serverConnectionCount.incrementAndGet()
+  private val PARAMETERS_NOT_AFFECT_SHARE_CONNECTION = Set(
+    // If it is SAVE, SC added overwrite internally, refer to DefaultSource.createRelation()
+    "overwrite",
+    Parameters.PARAM_POSTACTIONS,
+    Parameters.PARAM_PREACTIONS,
+    Parameters.PARAM_SF_QUERY,
+    Parameters.PARAM_SF_DBTABLE
+  )
+  private val connectionSharableCheckOptions =
+    parameters.parameters -- PARAMETERS_NOT_AFFECT_SHARE_CONNECTION
+
+  override def toString: String = "Connection Cache Key"
+
+  override def equals(obj: Any): Boolean = obj match {
+    case other: ConnectionCacheKey =>
+      connectionSharableCheckOptions.equals(other.connectionSharableCheckOptions)
+    case _ => false
+  }
+
+  override def hashCode(): Int =
```

Comment:
We uses a hash map:  to cache the JDBC connection:
`cachedJdbcConnections = mutable.Map[ConnectionCacheKey, Connection]()`

So, override `equals()` and `hashCode()` accordingy.
For `hashCode()`, we order the entries in `connectionSharableCheckOptions`, and make a string, and use the result String's hash code.

Reply:
Why not session ID?

Reply:
`ConnectionCacheKey` is created from `MergedParameters`. Two JDBC connections created for the same `MergedParameters` must has different session id.

Reply:
No matter what parameters are, two JDBC connections have different session ID, even they have same connection parameters.
My idea is, we can reuse Snowflake's connection ID, instead of generating a new one.